cartocrow/isoline_simplification/isoline_simplifier.cpp

```diff
@@ -0,0 +1,1511 @@
+
+			// Remove from Delaunay
+			delaunay_remove_e(edge);
+			delaunay_remove_e(st);
+			if (!delaunay_remove_p(t)) {
+				success = false;
+				break;
+			}
+			delaunay_remove_e(uv);
+			if (!delaunay_remove_p(u)) {
+				success = false;
+				break;
+			}
+		}
+		if (success) break;
+		++iters;
+	}
+
+	for (int i = 0; i < ladder.m_rungs.size(); i++) {
+		auto edge = ladder.m_rungs.at(i);
+		auto new_point = ladder.m_collapsed.at(i);
+
+		bool reversed = m_p_next.contains(edge.target()) && m_p_next.at(edge.target()) == edge.source();
+		const Point<K>& t = reversed ? edge.target() : edge.source();
+		const Point<K>& u = reversed ? edge.source() : edge.target();
+		assert(m_p_next.at(t) == u && m_p_prev.at(u) == t);
+		const Point<K>& s = m_p_prev.at(t);
+		const Point<K>& v = m_p_next.at(u);
+
+		// Insert into Delaunay
+		delaunay_insert_e(s, new_point, m_p_vertex.at(s));
+		auto seg_handle = delaunay_insert_e(new_point, v, m_p_vertex.at(v));
+		delaunay_insert_p(new_point, seg_handle);
+	}
+
+	// Update the rest
+	for (int i = 0; i < ladder.m_rungs.size(); i++) {
+		--m_current_complexity;
+
+		auto edge = ladder.m_rungs.at(i);
+		auto new_point = ladder.m_collapsed.at(i);
+
+		bool reversed = m_p_next.contains(edge.target()) && m_p_next.at(edge.target()) == edge.source();
+		const Point<K> t = reversed ? edge.target() : edge.source();
+		const Point<K> u = reversed ? edge.source() : edge.target();
+		assert(m_p_next.at(t) == u && m_p_prev.at(u) == t);
+		const Point<K> s = m_p_prev.at(t);
+		const Point<K> v = m_p_next.at(u);
+		const Segment<K> st = Segment<K>(s, t);
+		const Segment<K> uv = Segment<K>(u, v);
+
+		auto t_it = m_p_iterator.at(t);
+		auto u_it = m_p_iterator.at(u);
+		Isoline<K>* t_iso = m_p_isoline.at(t);
+		Isoline<K>* u_iso = m_p_isoline.at(u);
+		assert(t_iso == u_iso);
+
+		// Remove points from isolines
+		auto new_it = t_iso->m_points.insert(u_it, new_point);
+		t_iso->m_points.erase(t_it);
+		u_iso->m_points.erase(u_it);
+
+		auto remove_ladder_p = [this](Point<K> point) {
+			if (m_p_ladder.contains(point)) {
+				for (const auto& l : m_p_ladder.at(point)) {
+					l->m_old = true;
+				}
+				m_p_ladder.erase(point);
+			}
+		};
+
+		// Update some ladder info
+		remove_ladder_e(st);
+		remove_ladder_e(uv);
+		remove_ladder_e(st.opposite());
+		remove_ladder_e(uv.opposite());
+		remove_ladder_e(edge);
+		remove_ladder_e(edge.opposite());
+
+		remove_ladder_p(t);
+		remove_ladder_p(u);
+
+		// Update m_p_iterator
+		m_p_iterator.erase(t);
+		m_p_iterator.erase(u);
+		m_p_iterator[new_point] = new_it;
+
+		// Update m_p_isoline
+		m_p_isoline.erase(t);
+		m_p_isoline.erase(u);
+		m_p_isoline[new_point] = t_iso;
+
+		// Update prev and next
+		m_p_prev.erase(t);
+		m_p_next.erase(t);
+		m_p_prev.erase(u);
+		m_p_next.erase(u);
+		m_p_prev[v] = new_point;
+		m_p_next[s] = new_point;
+		if (m_p_prev.contains(new_point)) {
+			std::cerr << "Collapsed to existing point!!!" << std::endl;
```

Comment:
That's a lot of exclamation marks 😛

Reply:
I removed two of them :). Should never occur though for correct input.